### ISSUE TITLE
Allow setting ecmaVersion to latest

### ISSUE THREAD

**Issue by SimenB** [eslint/eslint#7528]
**What version are you using?**
3.9.1

**Problem I want to solve**
I want to be able to always use the latest available ecmaVersion

**Proposed solution**
Allow `'latest'` as a config option for `ecmaVersion`, aliasing whatever the latest version the included acorn supports. Kinda like https://babeljs.io/docs/plugins/preset-latest/

**Comment by platinumazure** [comment]
This would require changes upstream (to espree, possibly Acorn).

To me, this is similar to our `eslint:all` configuration: It's potentially dangerous (if the latest ecmaVersion involves backward-incompatible changes, those would be consumed without a major version bump), but if users choose to opt in, we could decide to wash our hands of semver responsibility for that particular use case.

Being a core change, this will require TSC approval.

**Comment by not-an-aardvark** [comment]
I don't think we should worry too much about backwards-incompatible changes, since JavaScript never really has any. When old code is executed in a newer browser, it's effectively using `ecmaVersion: latest` at runtime anyway.

**Comment by SimenB** [comment]
I did `require('acorn')` (thinking to provide a PR for this), but they don't currently export latest version (just default config, which is currently 7, not the latest 8). So might require upstream changes, yeah.

And I agree breaking changes in parsing shouldn't matter here, both because it's explicitly opt-in, and there aren't many, if any, breaking changes.

**Comment by kaicataldo** [comment]
I don't think comparing this enhancement to the babel-preset-latest is a good argument for its inclusion since static analysis is very different from transpilation. I'm personally not a fan of this idea and am not entirely sure what the benefit of doing this is, but here are a few thoughts.

* What does this mean for the `experimentalObjectRestSpread`? Would users expect this to be turned on with `latest`?
* Adding a new ecmaVersion in Espree is a minor version bump, so the actual latest version available to the user would be tied to Espree's release cycle, *not* ESLint's. This probably could be circumvented by having ESLint calculate what `latest` means, rather than having it be an upstream change.

**Comment by not-an-aardvark** [comment]
I would say that experimental features such as `experimentalObjectRestSpread` should be off by default; only officially accepted (stage 4) proposals should be included.

**Comment by kaicataldo** [comment]
Actually, ESLint "normalizes" version numbers but actually has no knowledge of the available options for `ecmVersion`, so to do what I suggested in the second bullet point we'd have to keep a list of supported `ecmaVersion`s.

**Comment by platinumazure** [comment]
Or we could augment espree with an API to get an array of supported
versions.

**Comment by ilyavolodin** [comment]
I don't really see a point in adding something like this. We don't trickle in small language enhancements into Espree one at a time. We usually go for a big-bang approach and release new major version of Espree with support for new language. If users will go through the trouble upgrading ESLint and Espree, they might as well modify one value in their config file to start using latest language features. This seems like a micro-optimization that requires a lot of thought and effort to solve the issue that takes less than 1 second to fix by hands.

**Comment by sindresorhus** [comment]
@ilyavolodin I always want the latest syntax. Browsers have versionless JS. I don't see why ESLint shouldn't at least have the ability to opt into such. IMHO `latest` should have been the default.

> This seems like a micro-optimization that requires a lot of thought and effort to solve the issue that takes less than 1 second to fix by hands.

Micro-optimization to you, better UX for many others. Doesn't seem like a lot of effort. Easiest solution would just be to hard-code a list of supported ECMAScript versions. It only needs updating once a year and then you're in control what `latest` means for ESLint.

1 second * ESLint users * configs, adds up.

**Comment by SimenB** [comment]
> Actually, ESLint "normalizes" version numbers but has no knowledge of the available options for `ecmVersion`, so to do what I suggested in the second bullet point we'd have to keep a list of supported `ecmaVersion`s.

You already maintain a list of all supported versions, albeit in docs and not in code.

http://eslint.org/docs/user-guide/configuring#specifying-parser-options

> `ecmaVersion` - set to 3, 5 (default), 6, 7, or 8 to specify the version of ECMAScript syntax you want to use. You can also set to 2015 (same as 6), 2016 (same as 7), or 2017 (same as 8) to use the year-based naming.

When that piece of documentation is updated, I don't think it's too much extra work to also update an array in the code, is it?

**Comment by kaicataldo** [comment]
> You already maintain a list of all supported versions, albeit in docs and not in code.
That's right. I wasn't suggesting it's a blocker for this enhancement - just laying out the current situation.

My concern is with `latest` being tied to Espree's release cycle rather than ESLint's, as I don't think this should change when ESLint hasn't had a new release. If `latest` is set in each ESLint version, I'm not against the addition of this.

**Comment by ljharb** [comment]
I like the idea of using "latest" - but if "latest" is supported, then adding a new ecmaVersion should change from semver-minor to semver-major.

**Comment by not-an-aardvark** [comment]
Could you clarify why it would be necessary to do a semver-major bump when adding a new ecmaVersion?

**Comment by nzakas** [comment]
I think we're getting a bit too far down the implementation rat hole with the current line of discussion, so let's pop the stack a bit.

To start, I'm on the fence about this feature. I can see some people wanting it, and I can also see why team members have concerns. That said, there is an easy way to implement this is a non-breaking manner.

First, we can just hardcode `"latest"` to be equal to some version directly in ESLint. We wouldn't need or want Espree to export its latest version, necessarily. There's less of a use case for Espree doing this than there is for ESLint.

So we hardcode some value for `"latest"` in the ESLint source code. That still leaves us free to add new versions to Espree without fear of breaking ESLint users. Then, the next time we do a major release for ESLint, we can update the the definition of `"latest"`. (To @ljharb's point, changing what `"latest"` means is best represented as a breaking change.)

In this case, `"latest"` ends up meaning, the "the latest ECMAScript version as of the last major release of ESLint" rather than "the latest ECMAScript version as of the last Espree release."

**Comment by ilyavolodin** [comment]
@nzakas Sorry, I'm not sure I understand. Are you saying we will create `latest` and manually change what constitutes "latest" with every major release? If that's the case, I'd be fine with that, but I'm not sure that's what's being asked here. We do major releases about once a year, if Espree starts supporting next version of ECMAScript a month after we did major ESLint release, and we will wait for at least a few month to do another major release, that sort of defeats the purpose of "latest", as far as I can see.

**Comment by SimenB** [comment]
Yes, my wish is for a way to always have ESLint parse whatever the newest version it supports, if I have to wait for an ESLint major (or bump it myself), I dont think much is gained.

Node doesn't consider adding language features a breaking change, even though the parsing might change. Why should ESLint?

**Comment by ljharb** [comment]
@SimenB because "it doesn't parse code i don't transpile" is a feature. if a non-major eslint update causes broken code in production, because i was relying on the linter failure to protect against well-meaning devs, then that's a breaking change.

**Comment by not-an-aardvark** [comment]
I disagree that *not* reporting a parse failure after having previously reported a failure would be a breaking change. This is basically the definition of a semver-minor change; it allows the user to use new functionality that would have caused an error in the previous version.

By that logic, any upgrade to Node.js that introduces new ES features is a breaking change, because one could write code that works on the new version but would cause an unexpected error if production is running on an earlier version.

**Comment by ljharb** [comment]
@not-an-aardvark the difference is that eslint is a verification tool, node and browsers are not. I'm using eslint to verify that code will run on my target environments - if it suddenly allows code that will not run, it's not serving one of its primary purposes.

**Comment by not-an-aardvark** [comment]
But if you have specific environments you're targetting, I don't think it would be a good idea to set `ecmaVersion: latest` anyway. I think by using that option, you would be opting into allowing the latest stage 4 features when they become available.

**Comment by ilyavolodin** [comment]
@not-an-aardvark You are making an assumption that new features will always result in *less* errors. Look at ES7 and `use strict` inside function with non-trivial arguments. That would result in *more* errors. Since proposal is to automate behavior of upgrading parser version, we can't predict if there will be less or more errors, as such we need to limit it to major releases only.

**Comment by platinumazure** [comment]
Does anyone have any thoughts about what I brought up earlier?

Specifically: We could do this (and tie it to the espree release, not that of ESLint) as long as this is documented as "voids your semver warranty", similar to the "eslint:all" configuration. So in other words, users can opt in at their own risk.

**Comment by SimenB** [comment]
@platinumazure that would be perfect for me

**Comment by ljharb** [comment]
The proper opt-in for "voids your semver warranty" is putting "*" in package.json.

**Comment by kaicataldo** [comment]
I'm still very much against tying this to Espree's release cycle, even with a disclaimer, for a few reasons:
 * I feel like just *having* this option encourages users to use it. It sounds very appealing to set the version once and never have to worry about it again. However, there is always a delay between when Espree lands the latest version and when ESLint fully supports the new features, because rules need to be updated for the new syntax. This means that there's a very high propbably that ESLint is broken for those users for that duration. Example: see the [following changes that had to be made to ensure `async`/`await` worked for all of ESLint's core rules](https://github.com/eslint/eslint/issues/7101). You could argue that the user doesn't have to use the new features - but then what's the point of `latest`? I'm not a fan of knowingly giving users an option that has a high probability of breaking their build.
 * Tying `latest` to Espree's release cycle means that two users (or even two builds by the same person) running the same version of ESLint could potentially have been linted with the parser set to two different `ecmaVersion`s. How can one have confidence in a quality assurance check if it's not deterministic?

The "eslint:all" configuration is different because it only changes on new releases. To be clear: I'm not against the inclusion of this option, but I do think it should be tied to ESLint's release cycle.

**Comment by platinumazure** [comment]
@kaicataldo has convinced me of why ecmaVersion: latest should not be determined solely by espree (I had forgotten about the async/await taking more than one minor release to fully implement in ESLint). I now agree that it should be tied to ESLint's release cycle if we're going to do this at all.

**Comment by nzakas** [comment]
Sorry, was laid up for quite a few days and only getting back to this now.

From the ESLint side, it looks like we're all in agreement that if we were to make a change, the value of `"latest"` would have to be tied to ESLint release cycles and not Espree release cycles. I think that's fine, and would basically be similar to use updating the default ecmaVersion with each major release.

If that doesn't solve the use case for @SimenB, then we probably don't have a path forward on this.

**Comment by SimenB** [comment]
If it's behind a major version, it doesn't really solve what I want. If it's bumped in a minor though (like version 8 was made available in 3.6 now), then that's perfectly fine with me!

**Comment by platinumazure** [comment]
I'd be :+1: to updating it in a minor ESLint version, since it means we at least control when *we think* ESLint is fully ready for any new syntax available in the new ecmaVersion, without unnecessarily holding it up in a major release. (In that case, of course, "ecmaVersion: latest" is somewhat "buyer beware" just like the "eslint:all" configuration.)

**Comment by nzakas** [comment]
I don't think we can make such a change in a minor version. I'll throw this on the TSC agenda.

TSC Summary: The proposal is to add `ecmaVersion: latest`, which would always point to the highest available ECMAScript version ESLint supports. The team does not feel that ESLint should automatically pull in the highest Espree ECMAScript version supported because an Espree release could opaquely affect existing ESLint versions. The team feels that you could hardcode `"latest"` to a specific ECMAScript version and update it periodically. Some think this update must be done in a major version, others suggest doing so in a minor version.

TSC Questions: 1) Are we in favor of having `ecmaVersion: latest` and 2) if yes, would we want to update the value of `latest` in a major version or a minor version?

**Comment by btmills** [comment]
After discussion in the [2016-11-10 TSC meeting](https://gitter.im/eslint/tsc-meetings/archives/2016/11/10), the team decided not to add this feature.

**Comment by sindresorhus** [comment]
Can you add a summary of the reasoning that led up to that decision? I don't have time to read meeting notes and I think it should be here for posterity.

**Comment by not-an-aardvark** [comment]
Here's the relevant section of the notes:

```
Nicholas C. Zakas @nzakas
Next is eslint/eslint#7528
TSC Summary: The proposal is to add ecmaVersion: latest, which would always point to the highest available ECMAScript version ESLint supports. The team does not feel that ESLint should automatically pull in the highest Espree ECMAScript version supported because an Espree release could opaquely affect existing ESLint versions. The team feels that you could hardcode "latest" to a specific ECMAScript version and update it periodically. Some think this update must be done in a major version, others suggest doing so in a minor version.
TSC Questions: 1) Are we in favor of having ecmaVersion: latest and 2) if yes, would we want to update the value of latest in a major version or a minor version?

alberto @alberto Nov 10 16:37
:+1: major

Nicholas C. Zakas @nzakas Nov 10 16:37
@alberto so you are :+1: for the feature and :+1: for changes in a major version?

alberto @alberto Nov 10 16:37
yes

Nicholas C. Zakas @nzakas Nov 10 16:37
I'm ambivalent about the feature, but if we did it, I'd want to update only in a major version.
Other thoughts?

Gyandeep Singh @gyandeeps Nov 10 16:38
then we have to do major upgrade every year?

Kai Cataldo @kaicataldo Nov 10 16:38
@nzakas I feel the same way as you. I don't really see a lot of value in the feature, but it doesn't seem harmful if it's in a major version. That being said, the user requesting the feature said that would defeat the purpose of their suggestion

Nicholas C. Zakas @nzakas Nov 10 16:39
@gyandeeps yes, which I think is something we'd end up doing anyway

Ilya Volodin @ilyavolodin Nov 10 16:39
I don't see a lot of value in the feature in general, but if we go with major version, I don't see any value in it at all

Brandon Mills @btmills Nov 10 16:39
Agree with @nzakas. The feature sounds dangerous, but I don't have to prevent others from living on the edge. Definitely a major though.

Toru Nagashima @mysticatea Nov 10 16:39
I'm :-1: for the feature, but if we accepted the feature, I think updating latest needs a major version.

Nicholas C. Zakas @nzakas Nov 10 16:40
okay, it looks like consensus is that we don't like the feature. Does anyone want to argue for it?
All right, we've resolve to not accept this feature.
```

(disclaimer: I'm not on the TSC)

tl;dr: It looks like there was agreement that updating `latest` would require a major version bump. But we don't want to have to do major releases too often, so `latest` would get out of date, which would defeat the purpose of the setting. Since there wouldn't really be a point in having an option called `latest` that really means "latest as of the last major ESLint release", we decided that it wasn't worth adding the setting at all.
